File: src/backend/internal/model/knowledge.go
package model

import (
	"encoding/json"
	"time"
)

// KnowledgeBase 知识库模型
type KnowledgeBase struct {
	ID          string          `json:"id" db:"id"`
	UserID      string          `json:"user_id" db:"user_id"`
	Name        string          `json:"name" db:"name"`
	Description string          `json:"description" db:"description"`
	Visibility  string          `json:"visibility" db:"visibility"`
	CreatedAt   time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time       `json:"updated_at" db:"updated_at"`
	Files       []KnowledgeFile `json:"files" db:"-"`
	FileCount   int             `json:"file_count" db:"file_count"`
	Username    string          `json:"username,omitempty" db:"username"`
}

// KnowledgeFile 知识库文件模型
type KnowledgeFile struct {
	ID              string     `json:"id" db:"id"`
	KnowledgeBaseID string     `json:"knowledge_base_id" db:"knowledge_base_id"`
	Filename        string     `json:"filename" db:"filename"`
	FilePath        string     `json:"-" db:"file_path"`
	URL             string     `json:"url,omitempty" db:"-"`
	FileSize        int64      `json:"size" db:"file_size"`
	MimeType        string     `json:"mime_type" db:"mime_type"`
	PreviewText     string     `json:"preview_text,omitempty" db:"preview_text"`
	PreviewType     string     `json:"preview_type" db:"preview_type"`
	RAGStatus       string     `json:"rag_status" db:"rag_status"`
	RAGError        string     `json:"rag_error,omitempty" db:"rag_error"`
	ChunkCount      int        `json:"chunk_count" db:"chunk_count"`
	IndexedAt       *time.Time `json:"indexed_at,omitempty" db:"indexed_at"`
	CreatedAt       time.Time  `json:"uploaded_at" db:"created_at"`
}

// KnowledgeChunk 是可独立召回并能追溯到原文件的知识片段。
type KnowledgeChunk struct {
	ID              string          `json:"id" db:"id"`
	KnowledgeBaseID string          `json:"knowledge_base_id" db:"knowledge_base_id"`
	FileID          string          `json:"file_id" db:"file_id"`
	ChunkIndex      int             `json:"chunk_index" db:"chunk_index"`
	ChunkType       string          `json:"chunk_type" db:"chunk_type"`
	Content         string          `json:"content" db:"content"`
	Summary         string          `json:"summary" db:"summary"`
	TokenCount      int             `json:"token_count" db:"token_count"`
	CharCount       int             `json:"char_count" db:"char_count"`
	Metadata        json.RawMessage `json:"metadata" db:"metadata"`
	Embedding       []float32       `json:"-" db:"-"`
	EmbeddingModel  string          `json:"embedding_model" db:"embedding_model"`
	ContentHash     string          `json:"content_hash" db:"content_hash"`
	CreatedAt       time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt       time.Time       `json:"updated_at" db:"updated_at"`
}

// KnowledgeChunkSearchResult 在分块内容之外返回 cosine 相似度和来源文件名。
type KnowledgeChunkSearchResult struct {
	KnowledgeChunk
	Filename string  `json:"filename" db:"filename"`
	Score    float64 `json:"score" db:"score"`
}

File: src/frontend/src/types/knowledge.ts
/** 知识库可见性 */
export type KnowledgeVisibility = 'private' | 'public';

export type KnowledgeRAGStatus = 'pending' | 'indexing' | 'indexed' | 'failed' | 'skipped';

/** 知识库中的文件 */
export interface KnowledgeFile {
  id: string;
  filename: string;
  url?: string;
  size: number;
  mime_type: string;
  preview_type: 'text' | 'image' | 'binary' | 'too_large';
  preview_text?: string;
  rag_status?: KnowledgeRAGStatus;
  rag_error?: string;
  chunk_count?: number;
  indexed_at?: string;
  uploaded_at: string;
}

export type KnowledgeRetrievalMode = 'keyword' | 'vector' | 'reranked';

export interface KnowledgeRAGResult {
  id: string;
  knowledge_base_id: string;
  file_id: string;
  filename: string;
  chunk_index: number;
  chunk_type: string;
  summary: string;
  content: string;
  score: number;
  vector_score: number;
  rerank_score?: number;
  embedding_model?: string;
  retrieval_mode: KnowledgeRetrievalMode;
}

export interface KnowledgeReindexResult {
  knowledge_base_id: string;
  total_files: number;
  indexed_files: number;
  failed_files: number;
  skipped_files: number;
}

export interface KnowledgeFileText {
  file_id: string;
  filename: string;
  mime_type: string;
  preview_type: KnowledgeFile['preview_type'];
  text: string;
  truncated: boolean;
}

/** 知识库 */
export interface KnowledgeBase {
  id: string;
  name: string;
  description?: string;
  visibility: KnowledgeVisibility;
  files: KnowledgeFile[];
  file_count: number;
  created_at: string;
  updated_at: string;
}

/** 群组中可用的知识库（含 owner 信息） */
export interface GroupKnowledgeBase {
  id: string;
  name: string;
  description?: string;
  visibility: KnowledgeVisibility;
  username: string;
  file_count: number;
  created_at: string;
  updated_at: string;
}
export interface CreateKnowledgeBaseRequest {
  name: string;
  description?: string;
  visibility?: KnowledgeVisibility;
}

/** 更新知识库请求 */
export interface UpdateKnowledgeBaseRequest {
  name?: string;
  description?: string;
  visibility?: KnowledgeVisibility;
}
